File: 0048-rotate-image/0048-rotate-image.java
class Solution {
    public void rotate(int[][] matrix) {

        int n = matrix.length;

        // STEP 1 : Transpose
        for(int i = 0; i < n - 1; i++) {
            for(int j = i + 1; j < n; j++) {

                int temp = matrix[i][j];
                matrix[i][j] = matrix[j][i];
                matrix[j][i] = temp;

            }
        }

        // STEP 2 : Reverse every row
        for(int i = 0; i < n; i++) {
            reverse(matrix[i]);
        }
    }

    public void reverse(int[] arr) {

        int left = 0;
        int right = arr.length - 1;

        while(left < right) {

            int temp = arr[left];
            arr[left] = arr[right];
            arr[right] = temp;

            left++;
            right--;
        }
    }
}

File: 0048-rotate-image/0048-rotate-image.cpp
class Solution {
public:
    void rotate(vector<vector<int>>& matrix) {

        int n = matrix.size();
        // vector<vector<int>> ans(n, vector<int>(n));

        // for(int i = 0; i < n; i++) {
        //     for(int j = 0; j < n; j++) {
        //         ans[j][n-1-i] = matrix[i][j];
        //     }
        // }

        // matrix = ans;

        /* OPTIMIAL SOL. STEP 1  =  TRANSPOSE THE MATRIX */
        for(int i = 0; i < n-1; i++) {
            for(int j = i+1; j < n; j++) {
                swap(matrix[i][j], matrix[j][i]);

            }
        }

        /* STEP 2  =  REVERSE EVERY ROW */
        for(int i = 0; i < n; i++) {
            // matrix[i]
            reverse(matrix[i].begin(), matrix[i].end());
        }
    }
};